File: src/main/java/me/pljr/pljrapispigot/database/DataSource.java
package me.pljr.pljrapispigot.database;

import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;
import me.pljr.pljrapispigot.managers.ConfigManager;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public class DataSource {
    private final String host;
    private final String port;
    private final String database;
    private final String username;
    private final String password;
    private final int maximumPoolSize;
    private final int maximumIdle;
    private final int maxLifetime;
    private final int connectionTimeout;
    private final boolean cachePrepStmts;
    private final int prepStmtCacheSize;
    private final int prepStmtCacheSqlLimit;

    private final HikariConfig config = new HikariConfig();
    private HikariDataSource ds;

    public DataSource(ConfigManager config) {
        host = config.getString("mysql.host");
        port = config.getString("mysql.port");
        database = config.getString("mysql.database");
        username = config.getString("mysql.username");
        password = config.getString("mysql.password");
        maximumPoolSize = config.getInt("mysql-settings.maximumPoolSize");
        maximumIdle = config.getInt("mysql-settings.maximumIdle");
        maxLifetime = config.getInt("mysql-settings.maxLifetime");
        connectionTimeout = config.getInt("mysql-settings.connectionTimeout");
        cachePrepStmts = config.getBoolean("mysql-settings.cachePrepStmts");
        prepStmtCacheSize = config.getInt("mysql-settings.prepStmtCacheSize");
        prepStmtCacheSqlLimit = config.getInt("mysql-settings.prepStmtCacheSqlLimit");
    }

    public void initPool(String name) {
        if (this.ds.isRunning()) return;
        this.config.setJdbcUrl("jdbc:mysql://" + this.host + ":" + Integer.parseInt(this.port) + "/" + this.database + "?characterEncoding=UTF-8&autoReconnect=true&useSSL=false");
        this.config.setUsername(this.username);
        this.config.setPassword(this.password);
        this.config.addDataSourceProperty("cachePrepStmts", this.cachePrepStmts);
        this.config.addDataSourceProperty("prepStmtCacheSize", this.prepStmtCacheSize);
        this.config.addDataSourceProperty("prepStmtCacheSqlLimit", this.prepStmtCacheSqlLimit);
        this.config.setMaximumPoolSize(this.maximumPoolSize);
        this.config.setMinimumIdle(this.maximumIdle);
        this.config.setMaxLifetime(this.maxLifetime);
        this.config.setConnectionTimeout(this.connectionTimeout);
        this.config.setPoolName(name);
        this.ds = new HikariDataSource(this.config);
    }

    public Connection getConnection() throws SQLException {
        return ds.getConnection();
    }

    public void close(Connection conn, PreparedStatement ps, ResultSet res) {
        if (conn != null) try {
            conn.close();
        } catch (SQLException ignored) {
        }
        if (ps != null) try {
            ps.close();
        } catch (SQLException ignored) {
        }
        if (res != null) try {
            res.close();
        } catch (SQLException ignored) {
        }
    }
}

File: src/main/kotlin/me/pljr/pljrapispigot/database/DataSource.kt
package me.pljr.pljrapispigot.database

import com.zaxxer.hikari.HikariConfig
import com.zaxxer.hikari.HikariDataSource
import me.pljr.pljrapispigot.config.ConfigManager
import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.SQLException

private const val CFG_PATH = "mysql"
private const val CFG_PATH_SETTINGS = "mysql-settings"

class DataSource(config: ConfigManager) {
    val hikariCfg = HikariConfig()
    var ds: HikariDataSource? = null

    val host = config.getString("$CFG_PATH.host")
    val port = config.getString("$CFG_PATH.port")
    val database = config.getString("$CFG_PATH.database")
    val username = config.getString("$CFG_PATH.username")
    val password = config.getString("$CFG_PATH.password")

    val maximumPoolSize = config.getInt("$CFG_PATH_SETTINGS.maximumPoolSize")
    val maximumIdle = config.getInt("$CFG_PATH_SETTINGS.maximumIdle")
    val maxLifeTime = config.getLong("$CFG_PATH_SETTINGS.maxLifeTime")
    val connectionTimeout = config.getLong("$CFG_PATH_SETTINGS.connectionTimeout")
    val cachePrepStmts = config.getBoolean("$CFG_PATH_SETTINGS.cachePrepStmts")
    val prepStmtCacheSize = config.getInt("$CFG_PATH_SETTINGS.prepStmtCacheSize")
    val prepStmtCacheSqlLimit = config.getInt("$CFG_PATH_SETTINGS.prepStmtCacheSqlLimit")

    fun initPool(name: String) {
        if (this.ds != null && this.ds!!.isRunning) return
        this.hikariCfg.jdbcUrl = "jdbc:mysql://$host:$port/$database?characterEncoding=UTF-8&autoReconnect=true&useSSL=false"
        this.hikariCfg.username = this.username
        this.hikariCfg.password = this.password
        this.hikariCfg.addDataSourceProperty("cachePrepStmts", this.cachePrepStmts)
        this.hikariCfg.addDataSourceProperty("prepStmtCacheSize", this.prepStmtCacheSize)
        this.hikariCfg.addDataSourceProperty("prepStmtCacheSqlLimit", this.prepStmtCacheSqlLimit)
        this.hikariCfg.maximumPoolSize = this.maximumPoolSize
        this.hikariCfg.minimumIdle = this.maximumIdle
        this.hikariCfg.maxLifetime = this.maxLifeTime
        this.hikariCfg.connectionTimeout = this.connectionTimeout
        this.hikariCfg.poolName = name
        this.ds = HikariDataSource(this.hikariCfg)
    }

    fun updateStatement(sql: String) {
        val statement = getConnection().prepareStatement(sql)
        statement.executeUpdate()
        close(statement)
    }

    fun getConnection(): Connection {
        return ds!!.connection
    }

    fun close(ps: PreparedStatement?) =  close(getConnection(), ps)

    fun close(conn: Connection?, ps: PreparedStatement?) = close(conn, ps, null)

    fun close(conn: Connection?, ps: PreparedStatement?, res: ResultSet?) {
        if (conn != null) try {
            conn.close()
        } catch (ignored: SQLException) {
        }
        if (ps != null) try {
            ps.close()
        } catch (ignored: SQLException) {
        }
        if (res != null) try {
            res.close()
        } catch (ignored: SQLException) {
        }
    }
}